cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -242,54 +252,72 @@ private void processMetrics() {
           }
         }
 
+        if (nextMessageId != null) {
+          metaKeyMap.put(topicIdMetaKey, nextMessageId);
+          lastMessageId = nextMessageId;
+        }
+
         if (records.isEmpty()) {
           LOG.trace("No metrics record to process.");
           return;
         }
 
-        persistRecords(records);
-        records.clear();
-        lastMessageId = nextMessageId;
-        persistMessageId(lastMessageId);
+        // Ensure there's only one thread can persist records when the persistingFlag is false.
+        // The thread set persistingFlag to true when it starts to persist.
+        if (!persistingFlag.compareAndSet(false, true)) {
+          return;
+        }
+        Map<TopicIdMetaKey, byte[]> metaKeyMapCopy = new HashMap<>(metaKeyMap);
+        List<MetricValues> recordsCopy = new ArrayList<>(records.size());
```

Comment:
Try not to call `size()` on a concurrent queue. Better just copy to a linked list.